R/query_explain.R

```diff
@@ -0,0 +1,118 @@
+##' Explain how a query has or has not matched. This is experimental
+##' and the output will change. At the moment, it can tell you why a
+##' query matches, or if fails to match based on one of a number of
+##' `&&`-ed together clauses.
+##'
+##' @title Explain a query
+##'
+##' @inheritParams orderly_search
+##'
+##' @return An object of class `orderly_query_explain`, which can be
+##'   inspected (contents subject to change) and which has a print
+##'   method which will show a user-friendly summary of the query
+##'   result.
+##'
+##' @export
+orderly_query_explain <- function(..., parameters = NULL,
+                                  envir = parent.frame(),
+                                  options = NULL, root = NULL) {
+  root <- root_open(root, locate = TRUE, require_orderly = FALSE)
+  query <- as_orderly_query(...)
+  options <- as_orderly_search_options(options)
+  found <- orderly_search(query, parameters = parameters, envir = envir,
+                          options = options, root = root)
+  query_simplified <- query_simplify(query)
+  ret <- list(found = found,
+              n = length(na.omit(found)), # latest() returns NA
+              outer = query_simplified$outer,
+              parts = list())
+
+  for (name in names(query_simplified$parts)) {
+    expr <- query_simplified$parts[[name]]
+    found <- orderly_search(expr, parameters = parameters, envir = envir,
+                            options = options, root = root)
+    ret$parts[[name]] <- list(
+      name = name,
+      str = deparse_query(expr, NULL, NULL),
+      expr = expr,
+      n = length(found),
+      found = found)
+  }
+
+  ## In the case where we have exactly one 'n == 0' case we can try
+  ## and find more near misses, but this will be a follow-on PR. The
+  ## idea is that we will identify the offending bit like:
+  ##
+  ## > parameter:x == value
+  ##
+  ## Then we evaluate the rest of the query - all the bits without
+  ## this chunk to get the context. Then look to see how we can
+  ## permute the name of the parameter to get a near miss, or the
+  ## value to get a near miss.
```

Comment:
Yeah this sounds great, I wonder if also as an extension could use string distance on potential packet names? To give some suggestions for `name == "y"` if `"y"` doesn't find any

Reply:
yeah, I was expecting to reuse the "did you mean" bits that we have already